### PR TITLE
Support game INIs that match all regions

### PR DESCRIPTION
**Pull request by JosJuice**
There are a lot of game INIs. Too many, in my opinion. Because they're so tiny and many, the space they take up in total on my hard drive is 8.63 MiB, even though the files themselves only are 1.20 MiB in total. Also, when extracting or deleting a newly downloaded dev release, most of the time spent is because of all the game INI files. I thought about some ways to reduce the number of files, like placing all the game INIs in a single ZIP file, or de-duplicating INIs so that only one INI is needed per game instead of one per region. This PR is about the latter. It will reduce the problem, but not get rid of it completely. It should also save time when editing game INIs.

How does it work? Basically, instead of having `RSBE01.ini`, `RSBJ01.ini`, `RSBK01.ini` and `RSBP01.ini`, just a `RSB-01.ini` is enough. INIs without region wildcards will work like before, and they will still be needed for INIs that contain things like patches and AR codes. If both a file with a wildcard and a file without a wildcard match the current game, both will be used, and the content of the file without a wildcard will be prioritized if there is a conflict. Note that the code only supports wildcards for the fourth character, not any other characters.
EDIT: [I have changed how this works.](https://github.com/dolphin-emu/dolphin/pull/2018#issuecomment-73382847)

This PR doesn't de-duplicate any of the INIs (at least not yet). I could write a little program that detects duplicates and turns them into INIs with wildcards, but I'm wondering how people think the comments on top of the INIs should be handled. Should they contain all game IDs for the game? Should they contain the game ID with a wildcard? Should they contain no game ID at all?

### CODE REVIEW

**Comment by Tilka** [comment]
I like this idea. As for comments, I'd go for all game IDs because that way it's more user-friendly to search for them.

File: Source/Core/Core/CoreParameter.cpp [hunk 1]
```diff
@@ -402,28 +402,42 @@ void SCoreStartupParameter::CheckMemcardPath(std::string& memcardPath, std::stri
...
+IniFile SCoreStartupParameter::LoadGameIni() const
 {
 	IniFile game_ini;
-	game_ini.Load(m_strGameIniLocal);
+	LoadGameIni(&game_ini, File::GetSysDirectory() + GAMESETTINGS_DIR DIR_SEP);
+	LoadGameIni(&game_ini, File::GetUserPath(D_GAMESETTINGS_IDX));
 	return game_ini;
 }
+
+void SCoreStartupParameter::LoadGameIni(IniFile* game_ini, std::string path) const
```
Comment (lioncash) [inline Source/Core/Core/CoreParameter.cpp]: `path` can be a const reference here.

File: Source/Core/Core/CoreParameter.cpp [hunk 2]
```diff
@@ -402,28 +402,42 @@ void SCoreStartupParameter::CheckMemcardPath(std::string& memcardPath, std::stri
...
+		std::string id_without_region(GetUniqueID());
+		id_without_region.replace(3, 1, "-");
+		game_ini->Load(File::GetSysDirectory() + GAMESETTINGS_DIR DIR_SEP + id_without_region + ".ini", true);
+	}
+
+	// Regular INIs
+	game_ini->Load(path + GetUniqueID() + ".ini", true);
+
+	// INIs with specific revisions
+	if (m_strRevisionSpecificUniqueID != "")
```
Comment (lioncash) [inline Source/Core/Core/CoreParameter.cpp]: `if (!m_strRevisionSpecificUniqueID.empty())`

**Comment by Stevoisiak** [comment]
My god, this seems incredibly useful. Can someone test if starting Dolphin with a large gamelist is any slower/faster with this PR?

**Comment by JosJuice** [comment]
The game list doesn't use this code as far as I know, and it has no reason to.

**Comment by Stevoisiak** [comment]
IIRC, the game list uses this to display the emulation state rating, as well as any emulation notes. I'm pretty sure this wouldn't cause any noticeable differences in the startup time, but better safe than sorry.

**Comment by JosJuice** [comment]
Oh right, the emulation rating... I forgot about that. Sounds like something worth testing.

**Comment by Linktothepast** [comment]
@Stevoisiak Haven't tested but probably not since the chances of someone having the same games for different game regions is not that high. For example only if you had Twilight Princess for all game regions they would share the same gameini.
@JosJuice The info from the gameinis shows up in the gui, doesn't it? Is it saved in a separate file though in user cache?

**Comment by JosJuice** [comment]
Found the emulation state parsing code in GameListCtrl. It parses the INIs manually without going through BootManager, so I'll need to fix it to support region wildcards.

**Comment by Linktothepast** [comment]
Ok, my opinion is that for maintaining seems a bit more complicated, right now i can immediately find the gameini in the folder i want to edit by it's filename. The difference in the filenames is hindering easy access. Personally i would prefer if every gameini was merged in a single txt file for dealing with the increasing number of gameinis.

**Comment by JosJuice** [comment]
How are you accessing the game INIs? If they're sorted alphabetically, the INIs with region wildcards are in the same place as the ones without.

**Comment by Linktothepast** [comment]
Alphabetically, but often i just type search to find them.

**Comment by Stevoisiak** [comment]
The full INI names would be in the header comments, so you should still be able to search the full gameID without any issues.

**Comment by Linktothepast** [comment]
Hm and how will the differences be handled between game inis of different regions? Can you upload the result of automatically de - duplicate inis to test? Also what happens with locally saved gameinis, are they also affected by this change?

**Comment by Stevoisiak** [comment]
Linktothepast: The idea is that this will only go into effect if it finds a "Generalized" ini region file. All old INI files would continue working as normal.

**Comment by Linktothepast** [comment]
Oh so local gameinis would still be saved by the old naming and only if already renamed gameinis exist to the new format they would be read? Ok, now it doesn't sound that bad.

**Comment by JosJuice** [comment]
I haven't written a de-duplicating program yet, so I can't get any results, but what I'm planning to do is to not de-duplicate if there is any difference other than the game ID.

User game INIs will continue working as before. It will be possible to use region wildcards with them, but I don't think people actually are going to do that.

**Comment by Linktothepast** [comment]
Ok final question, is what i proposed of using a single txt file for all the database gameinis impossible? Something similar to what happens with pcsx2 where they use only a single database file for all games.

**Comment by JosJuice** [comment]
That's possible, but it would require bigger changes.

File: Source/Core/Core/CoreParameter.cpp [hunk 3]
```diff
@@ -402,28 +402,42 @@ void SCoreStartupParameter::CheckMemcardPath(std::string& memcardPath, std::stri
...
 }
+
+void SCoreStartupParameter::LoadGameIni(IniFile* game_ini, const std::string& path) const
+{
+	// INIs with region wildcards
+	if (GetUniqueID().size() >= 4)
+	{
+		std::string id_without_region(GetUniqueID());
+		id_without_region.replace(3, 1, "-");
+		game_ini->Load(File::GetSysDirectory() + GAMESETTINGS_DIR DIR_SEP + id_without_region + ".ini", true);
```
Comment (RachelBryk) [inline Source/Core/Core/CoreParameter.cpp]: "game_ini->Load(File::GetSysDirectory() + GAMESETTINGS_DIR DIR_SEP" should be "path"

**Comment by JosJuice** [comment]
All code that accesses game INIs is now updated, except for some code in watch and breakpoint windows that loads and saves user game INIs, because region wildcards would be more of an obstacle rather than something helpful there.

**Comment by JosJuice** [comment]
561 INIs have been merged. That's less than I thought would be possible, but it's still a nice number. 284 INIs couldn't be automatically merged, because they had differences in lines other than the first one. I suspect that some (or many?) of them could be merged manually.

**Comment by Linktothepast** [comment]
Hmm, btw will the use of wildcards cover cases where a region was missing before? For example let's say a virtual console game was missing the japanese region gameini before, using a wildcard instead will cover the japanese region too that was missing before from the database, correct? This will make the database more complete with less trouble.

**Comment by Linktothepast** [comment]
Anyway, if this is ready for merge, the sooner it is merged the better, to check manually for those not automatically merged cases and alter some ini changes i made accordingly to this pr. Otherwise there will be merge conflicts. Any gameini prs not merged by others should also be merged before this pr.

**Comment by JosJuice** [comment]
Yes, wildcards will cover regions that were missing before. Adding the additional game IDs to the comment at the top would be nice for the sake of documentation, but the only step that is required technically is to replace the fourth character with a `-`. In case there is interest in it, I could add wildcards to all INIs that only have one region right now. It would cause some false positives, but it would also give us better coverage without having to do anything manually.

If the last two characters are different, it isn't possible to cover both games with a single INI. Should I change that? It would just be three 3-5 extra lines of code or so.

**Comment by Linktothepast** [comment]
@JosJuice  Yes, i would like that.

**Comment by JosJuice** [comment]
Which thing did you mean? Wildcards for the last two characters?

**Comment by Linktothepast** [comment]
Yes.

**Comment by Linktothepast** [comment]
But i wouldn't mind if you put wildcards to single ini's too, normally it shouldn't cause false positives. It would be a pretty rare case.

**Comment by JosJuice** [comment]
I changed how things work. Instead of using a `-` as a wildcard, you now just make a filename that contains the three first characters, like `RSB.ini`. This will match all game IDs that begin with `RSB`, completely disregarding any characters after the three first ones. (Note that you must use exactly three characters. Any other number won't work.)

**Comment by Linktothepast** [comment]
You should probably remove your previous implementation inis and the redundant ones.

**Comment by JosJuice** [comment]
I've done that already.

**Comment by Linktothepast** [comment]
They all appear in the build provided by the buildbot.

**Comment by JosJuice** [comment]
Maybe there's a problem with the buildbot? I'm pretty sure I deleted them from the repo, because I did a hard reset on the old de-duplication commit, and I can't see any INIs with dashes locally.

**Comment by Stevoisiak** [comment]
I'm not sure I like the idea of just entering the first 3 letters of an INI file. I can easily see a system like that causing issues with homebrew and modded games. I know that most Brawl mods retain their RSBE ID codes. (Including Project M)

**Comment by JosJuice** [comment]
Don't mods generally need the same settings as the base game? Also, homebrew games are usually just DOLs/ELFs, so they don't have game IDs.

**Comment by Linktothepast** [comment]
@dolphin-emu-bot rebuild.

**Comment by Stevoisiak** [comment]
True, I suppose most mods would share settings, but I still don't like the idea of mods sharing the same emulation notes and ratings as the main games by default. I'd rather mods need to be rated manually.

Aisde from that, I suppose the only other issues would arise from Triforce games and custom Virtual Console titles. Triforce games in particular tend to have very strange IDs given to them. ([GGPE01](https://wiki.dolphin-emu.org/index.php?title=GGPE01) vs [GGPE02](https://wiki.dolphin-emu.org/index.php?title=GGPE02)), but I suppose we can handle those on a case by case basis.

**Comment by Linktothepast** [comment]
We don't have any mods in the gameini database though, the user could still change something locally if he finds a small difference like the rating.

**Comment by Stevoisiak** [comment]
@Linktothepast I'm saying if we did a match for "RSB---" instead of "RSBE01", then all of the emulation settings, ratings, and notes from the main game would be automatically be inherited by any mod using the same ID. A quick search on GameTDB shows that (at least a few games do this) [http://www.gametdb.com/Main/Results?cx=partner-pub-4620983010111854%3A3114910054&cof=FORID%3A11&ie=UTF-8&q=RSBE], (including Project M)[http://www.gametdb.com/Wii/RSBEPM].

**Comment by Linktothepast** [comment]
I know but i can't see the bad thing in that considering no gameinis at all existed before? If a user edited his gameini for a mod locally it would still have his changes, if he didn't he will at least have the settings for the original game. It is better than a complete lack of settings even if the gameini is slightly inaccurate.

**Comment by Stevoisiak** [comment]
I'm less worried about the setting being inherited and more worried that if a user sees that the emulation state for a mod is 5 stars, they assume that it will run perfectly when we haven't done any testing.

**Comment by Linktothepast** [comment]
We don't officially support mods though as far as i know, it's up to the user to deal with issues considering mods.

**Comment by Stevoisiak** [comment]
We don't support mods, which is why I don't want Dolphin giving mods a 5 star emulation rating simply because we know the game the mod was based on runs fine.

**Comment by JosJuice** [comment]
Are emulation state ratings something that users really care about? And wouldn't we have the same problem for all mods that use the same ID as the base game?

**Comment by Linktothepast** [comment]
What is the alternative though? Would you prefer missing gameinis of various game regions instead like before in order for the above regarding gamemods not to happen?

**Comment by Stevoisiak** [comment]
@Linktothepast You misunderstand me. I would LOVE to see this merged. I think using a wildcard for the region code is a brilliant idea. I have no objections against using `RSB-01.ini` as a general settings file for all regions of a game. I'm only worried about using a wildcard for the publisher ID.

It's entirely possible I'm just being picky about this, so if nobody else has any issues with ignoring the publisher ID, then I'm fine with it.

**Comment by Linktothepast** [comment]
There are plenty of games that have something similar to this though:
SVVEG9 = The Croods: Prehistoric Party!
SVVPAF = The Croods: Prehistoric Party!
And changing just the region code would have a missing gameini, that's why i suggested it to JosJuice. There would be missing game regions of the same game otherwise.

**Comment by pauldacheez** [comment]
There are some retail games with the same first three letters and different publisher IDs that have zero relation, e.g. GSXE8P (Sanic Advenchur DX) and GSXE64 (Star Wars: The ~~MMU Test~~ Clone Wars). I'd be more careful with the three-letter stuff, but the three-letter-plus-publisher-ID stuff is perfectly fine IMO.

**Comment by magumagu** [comment]
The Sonic game is GXSE8P.

**Comment by CarlKenner** [comment]
I like this idea, even though I use region-specific and revision-specific ini files for a few things. This is especially good for Japanese and Korean users because most Japanese and Korean games don't have INI files.

**Comment by Linktothepast** [comment]
If everything is ok please merge so that further work in ini files can be done.

**Comment by Linktothepast** [comment]
There are a lot of gameini changes that need to be done and i can't work with them if this isn't merged, it will cause a lot of unnecessary work for me plus i will have to redo a pr after this will be merged to fine-tune the ini files. Who do i need to annoy about it? @skidau , sorry for the annoyance but please merge if this pr is ok.

**Comment by Linktothepast** [comment]
Final appeal, @delroth there are a lot of gameinis that need to be updated due to the changes that are made recently, if this pr seems fine and it is going to be merged please do, otherwise if this change is dismissed it would still be good to know to plan the changes i will make accordingly.

**Comment by Linktothepast** [comment]
No answers, i take it this will not be merged then and i will change inis according to that.

**Comment by degasus** [comment]
@Linktothepast or more likely that nobody was motivated enough to look over this patch.... As our ini maintainer: Do you like this PR?

**Comment by Linktothepast** [comment]
Yes, it covers cases of regions that were previously missing from the database with no extra work involved, and you can still use region specific gameinis if necessary. Which is a win - win situation. No extra work needed for missing gameinis and things will become more tidy in the database. http://www.gametdb.com/ that i use is not entirely complete database i think regarding gameregions, there were times in the past that new game specific regions were added afterwards, and those cases will also be covered by this pr.

**Comment by Linktothepast** [comment]
Nice, thanks!

**Comment by Linktothepast** [comment]
Jocjuice, it seems you didn't delete the unnecessary inis though, can you upload the ones in your database to use as a base for the changes i will make, thanks.

**Comment by JosJuice** [comment]
I'm still pretty sure I deleted them... Anyway, I'm away over the weekend with only my phone, so I can't send any files until Monday.

**Comment by Linktothepast** [comment]
Damn, i will try to do it manually then. Anyway i will still need to manually check them all to finetune them to the new system, in case some weren't merged for a minor difference, etc.

**Comment by Linktothepast** [comment]
Found another small issue, with the unnecessary ini files removed, while things work fine, you can't open the general ini in the dolphin gui. It works but it doesn't show up to check things, you have to search for the ini file manually in the folder for details of it's contents.

**Comment by degasus** [comment]
@Linktothepast Which one should we open on this button? I mean it's easy to just open the other path, but it will be hard to decide what to do if there different files...

**Comment by Linktothepast** [comment]
If a file with the complete gameini id exists it should open, if one with just the 3 first letters exists it should open in it's place instead, which doesn't happen now.

**Comment by JosJuice** [comment]
@Linktothepast: How are you downloading the INIs? If it only a problem with the buildbot, GitHub's "Download ZIP" option and regular git fetches/pulls should work.

**Comment by Linktothepast** [comment]
Thanks JosJuice, it is definitely a buildbot issue. Did the above and the inis are correct.

**Comment by Stevoisiak** [comment]
@Linktothepast: I want to help work on converting our INI files to the new system, but I'd rather not cause any merge conflicts with your work.

**Comment by adamd10d** [comment]
While this isn't *exactly *the same:

GB*0*E51 - Burnout (1)
GB*4*E51 - Burnout 2: Point of Impact

**Comment by Linktothepast** [comment]
Stevoisiak thanks for the offer, i plan to do it on steps and post a gameini pr when some important prs get merged, you can take it from there if you wish.
Adamd.. the fourth to sixth letter are common and not the third

**Pull request merged**